language/vm/vm_runtime/vm_runtime_types/src/value.rs

```diff
@@ -50,7 +50,7 @@ mod vm_types;
 /// represented. In that sense instances of `[Value]` are what gets pushed/popped on the
 /// stack and effectively the values representable in Move.
 #[derive(PartialEq, Eq, Debug, Clone)]
-enum ValueImpl {
+pub(crate) enum ValueImpl {
```

Comment:
This change and the one below seem unfortunate. The other native `Vector` procedure implementations didn't need to change the visibility--perhaps we can mimic their approach?

Reply:
Yeah this is something I wasn't very sure. Planning to have a conversation with @dariorussi on this.

Reply:
Without `pub(crate)` the function above will need to be written like this:
```
    pub fn native_destroy_empty(mut args: VecDeque<Value>) -> NativeReturnStatus {
        if let Some(v) = args.pop_front() {
            if let Some(NativeStructValue::Vector(NativeVector(v))) =
                v.value_as::<NativeStructValue>()
            {
                return if v.is_empty() {
                    NativeReturnStatus::Success {
                        cost: DESTROY_EMPTY_VEC_COST,
                        return_values: vec![],
                    }
                } else {
                    NativeReturnStatus::Aborted {
                        cost: DESTROY_EMPTY_VEC_COST,
                        error_code: DESTROY_NON_EMPTY_VEC,
                    }
                };
            }
        }
        NativeReturnStatus::InvalidArguments
    }
```
Personally I found it cumbersome compared to the nested pattern matching approach, but understand we may have a good reason to make `value_impl` private.

Reply:
I agree that this is not as nice, but I think it's preferable to /safer than to opening up the internals of `ValueImpl`. It also ensures that the VM can change its internal representation of `ValueImpl` without breaking clients.

Reply:
Cool. I'll get rid the `pub(crate)` for now.

Reply:
yeah I think the way we wrote it above is more in line with how values are intended to be used and I don't think it'll be a big deal.
I wonder however if NativeStruct should have a more direct (inlined) representation in the Value story. I have not thought much about it and neither looked at the implications but it's kind of unfortunate that we have to enter a different "namespace" when it comes to native structures.
Please think about it a bit, though let's not have that block progress on this diff